### PR TITLE
Adding implementation for tree-map.

### PR DESCRIPTION
**Pull request by mdumbrava20**

### CODE REVIEW

File: src/main/java/com/endava/internship/collections/Node.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Setter @Getter
+public class Node {
+    private final Student key;
+    private Integer value;
+
+    private Node left = null;
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/Node.java]: Instance variables are null by default, not need to write "=null"

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 2]
```diff
@@ -1,78 +1,248 @@
...
+            return true;
+        }
+        if (current.getKey().compareTo((Student) key) > 0){
+            return findKey(current.getRight(), key);
+        }
+        else {
+            return findKey(current.getLeft(), key);
+        }
+    }
+    public boolean findValue(Node current, Object value, Boolean isPresent){
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentMap.java]: make this method private. 
This method does not return a value if it finds it, it just tells if it exists or not so rename it to something like "isValuePresent"

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 3]
```diff
@@ -1,78 +1,248 @@
...
+            }else {
+                return getNodeReference(current.getLeft(),current, key);
+            }
+        }else {
+            return null;
+        }
+    }
+    private Node clearTree(Node current){
+        if (current == null)
+            return null;
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentMap.java]: refactor this method, it returns null either way

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 4]
```diff
@@ -1,78 +1,248 @@
 package com.endava.internship.collections;
 
-import java.util.Collection;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
 
 public class StudentMap implements Map<Student, Integer> {
+    private int size = 0;
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentMap.java]: int is 0 by default

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 5]
```diff
@@ -1,78 +1,248 @@
...
-        return null;
-    }
-
-    @Override
-    public Integer put(Student student, Integer integer) {
-        //TODO
-        return null;
+        if (!containsKey(o))
+            return null;
+        else
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentMap.java]: refactor this to get rid of else

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 6]
```diff
@@ -1,78 +1,248 @@
...
+            return getValueByKey(head, o);
     }
 
     @Override
     public Integer remove(Object o) {
-        //TODO
-        return null;
+        if (!findKey(head,o))
+            return null;
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentMap.java]: same here

File: src/main/java/com/endava/internship/collections/Student.java [hunk 7]
```diff
@@ -1,11 +1,16 @@
...
 
 /**
  * The class that defines the element that will be contained by your collection
  */
-public class Student //TODO consider implementing any interfaces necessary for your collection
+@Getter @Setter
+public class Student implements Comparable<Student> //TODO consider implementing any interfaces necessary for your collection
 {
     private String name;
     private LocalDate dateOfBirth;
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/Student.java:18]: Before pushing your changes go to every file and press Ctrl + Alt + L to format the code

File: src/main/java/com/endava/internship/collections/Node.java (hunk 1 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/collections/Node.java]: Fixed.

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 2 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Fixed.

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 3 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Fixed.

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 4 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Fixed.

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 5 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Refactored.

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 6 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Refactored.

File: src/main/java/com/endava/internship/collections/Student.java (hunk 7 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/collections/Student.java:18]: Fixed.

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 8]
```diff
@@ -1,78 +1,247 @@
...
+        } else {
+            return null;
+        }
+    }
+
+    private Node clearTree(Node current) {
+        if (current == null) return null;
+        current.setLeft(clearTree(current.getLeft()));
+        current.setRight(clearTree(current.getRight()));
+        return current;
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentMap.java:214]: if (current != null) {
        current.setLeft(clearTree(current.getLeft()));
        current.setRight(clearTree(current.getRight()));
        }
        return current;

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 9]
```diff
@@ -1,78 +1,247 @@
 package com.endava.internship.collections;
 
-import java.util.Collection;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
 
 public class StudentMap implements Map<Student, Integer> {
+    private int size;
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentMap.java:6]: Format code please

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 8 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/collections/StudentMap.java:214]: Fixed.

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 9 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/collections/StudentMap.java:6]: Fixed.

**Review by vermurachi — APPROVED** [review]

**Pull request merged**